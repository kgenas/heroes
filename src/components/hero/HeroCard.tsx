import { Link } from 'react-router-dom';

//* para que no de error en require.context
//? npm i -D @types/webpack-env
const heroImages = require.context('../../assets',true);

interface valuePros{
    id:string;
    superhero:string;
    alter_ego:string;
    first_appearance:string;
    characters:string
}

export const HeroCard = ({
    id,
    superhero,
    alter_ego,
    first_appearance,
    characters
}:valuePros) =>{
    const imagePath = heroImages(`./${id}.jpg`);
    return(
        <div className="col animate__animated animate__fadeIn">
            <div className="card">
                <div className="row no-gutters">
                    <div className="col-4">
                        <img src={ imagePath } className ='card-img' alt = { superhero } />
                    </div>
                    <div className="col-8">
                        <div className="card-body">
                            <h5 className="card-litle">{ superhero }</h5>
                            <p className="card-test">{ alter_ego }</p>
                            {
                                (alter_ego !== characters )&&
                                <p className="text-muted">{ characters }</p>
                            }
                            <p className="card-text">
                                <small className="text-muted">{ first_appearance }</small>
                            </p>
                            <Link to={`/hero/${id}`}>
                                Mas...
                            </Link>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    ) 
}