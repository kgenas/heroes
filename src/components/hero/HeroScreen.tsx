import { useParams, useNavigate, Navigate } from 'react-router-dom';
import { useMemo } from 'react';
import { getHeroById } from '../../selectors/getHeroById';

export const HeroScreen = () =>{

    //* parametro de la ruta del navegador
    const { heroeId } = useParams();
    const navigate = useNavigate();
    interface valuePros{
        id:string;
        superhero:string;
        alter_ego:string;
        first_appearance:string;
        characters:string;
        publisher:string;
    }

    const hero = useMemo( () => getHeroById( heroeId ),[ heroeId ] )
    const handleReturn = () =>{
        navigate(-1);//*return previous page
    }
    if(!hero){
        return <Navigate to='/'/> //* return origin state page
    }

    const {
        id,
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    }:valuePros = hero;
    
    return(
        <div className='row mt-5'>
            <div className='col-4'>
                <img
                    src = { require(`../../assets/${id}.jpg`) }
                    alt = { superhero }
                    className ='img-thumbnail animate__animated animate__fadeInLeft'
                />
            </div>
            <div className='col-8'>
                    <h3>{ superhero }</h3>
                    <ul className='list-group list-group-flush'>
                        <li className='list-group-item'><b>Alter ego:</b>{ alter_ego }</li>
                        <li className='list-group-item'><b>Publisher:</b>{ publisher }</li>
                        <li className='list-group-item'><b>First Appearance:</b>{ first_appearance }</li>
                    </ul>
                    <h5 className='my-3'>Characters</h5>
                    <p>{ characters }</p>
                    <button
                        className='btn btn-outline-info'
                        onClick={ handleReturn }
                    >
                        Return
                    </button>
            </div> 
        </div>
    );
};