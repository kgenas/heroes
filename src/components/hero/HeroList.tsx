import { useMemo } from "react"
import { getHeroesAll } from "../../selectors/getHeroesAll";
import { getHeroesPublisher } from "../../selectors/getHeroesPublisher"
import { HeroCard } from "./HeroCard";

interface valueProps {
    publisher:string;
}
export const HeroList = ( {publisher}:valueProps ) =>{


    const heroes = useMemo( () => getHeroesPublisher( publisher ),[ publisher ]);
    return(
        <div className="row rows-cols-1 row-cols-md-3 g-3 animate__animated animate__fadeIn">
            {
                heroes.map( hero => (
                    <HeroCard key={ hero.id}{...hero}/>              
                ))
            }
        </div>
    );
}

export const HeroAll = () =>{


    const heroes = getHeroesAll();
    return(
        <div className="row rows-cols-1 row-cols-md-3 g-3 animate__animated animate__fadeIn">
            {
                heroes.map( hero => (
                    <HeroCard key={ hero.id}{...hero}/>              
                ))
            }
        </div>
    );
}