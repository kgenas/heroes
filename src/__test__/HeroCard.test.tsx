
import { HeroCard } from "../components/hero/HeroCard";
import { renderWithProviders } from "../utils/test-utils";
import { screen } from "@testing-library/react";
describe('HeroCard',() =>{
    test('Render HeroCard',() =>{
        renderWithProviders(<HeroCard        
            key={"dc-batman"}
            id={"dc-batman"}
            superhero={"Batman"}
            alter_ego={"Bruce Wayne"}
            first_appearance={"Detective Comics #27"}
            characters={"Bruce Wayne"}    
        />);
        //eslint-disable-next-line          
        screen.debug();
    })
    test('HeroCard Snapshot',() =>{
        const { baseElement } =
        renderWithProviders(<HeroCard        
            key={"dc-batman"}
            id={"dc-batman"}
            superhero={"Batman"}
            alter_ego={"Bruce Wayne"}
            first_appearance={"Detective Comics #27"}
            characters={"Bruce Wayne"}    
        />);
        expect( baseElement ).toMatchSnapshot();
    })
    test('click HeroCard',() =>{
        renderWithProviders(<HeroCard        
            key={"dc-batman"}
            id={"dc-batman"}
            superhero={"Batman"}
            alter_ego={"Bruce Wayne"}
            first_appearance={"Detective Comics #27"}
            characters={"Bruce Wayne"}    
        />);
        //eslint-disable-next-line          
        expect(screen.getByRole('link',{ name:'Mas...'})).toHaveAttribute('href','/hero/dc-batman');
    })
})