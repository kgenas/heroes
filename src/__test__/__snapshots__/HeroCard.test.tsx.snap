// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroCard HeroCard Snapshot 1`] = `
<body>
  <div>
    <div
      class="col animate__animated animate__fadeIn"
    >
      <div
        class="card"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Batman"
              class="card-img"
              src="dc-batman.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-litle"
              >
                Batman
              </h5>
              <p
                class="card-test"
              >
                Bruce Wayne
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
